### ISSUE TITLE
Tighten validation for `id` fields

### ISSUE THREAD

**Issue by amercader** [ckan/ckan#7604]
Having loose or inconsistent validation on `id` fields can lead to confusing behaviour or security concerns. Let's do a review of the various entities with `id` field in the model and apply the same validation to all of them (if possible). Validators should include:

* "Can only be changed by a sysadmin": custom ids only make sense in the context of specific scenarios like migrations, it make sense that only sysadmins (or plugins that ignore auth) touch them
* "Value is a valid UUID". In the past we've allowed different values here to support potential use cases but it's probably safer to enforce UUIDs for all ids and use custom fields if needed
* "Id does not already exist when creating": to prevent taking over other entities

Any that I missed?

Some of these are already applied to some entities, but it would be good to be consistent across all models
* Dataset
* Resource
* Resource View
* Groups (org/groups)
* User
* Activity
* Extras
* Tags (?)
* ...

**Comment by pdelboca** [comment]
@amercader enforcing `UUID` will have conflicts with the current implementation of `ckanext-harvest` right? 

Locally (after running a CKAN Harvester) I have resources with ID `turismo_f7ce925a-0a48-4e38-8f48-360d6513fd83` that will not validate under this changes.

The following is an example of `resource_show` with `?id=turismo_f7ce925a-0a48-4e38-8f48-360d6513fd83`:

```
{
  "help": "http://localhost:5000/api/3/action/help_show?name=resource_show",
  "success": true,
  "result": {
    "accessURL": "http://datos.yvera.gob.ar/dataset/fbc269ea-5f71-45b6-b70c-8eb38a03b8db/resource/f7ce925a-0a48-4e38-8f48-360d6513fd83",
    "attributesDescription": "[{\"type\": \"string\", \"description\": \"Temporada\", \"title\": \"temporada\"}, {\"type\": \"string\", \"description\": \"Puerto de amarre\", \"title\": \"puerto\"}, {\"type\": \"string\", \"description\": \"Nombre del crucero\", \"title\": \"crucero\"}, {\"specialTypeDetail\": \"cruceristas\", \"type\": \"number\", \"description\": \"Cantidad de cruceristas\", \"title\": \"cruceristas\"}, {\"type\": \"string\", \"description\": \"Especificaciones sobre los valores ausentes o nulos presentes en la serie de datos.\", \"title\": \"observaciones\"}]",
    "cache_last_updated": null,
    "cache_url": null,
    "created": "2023-09-29T14:43:19.101248",
    "datastore_active": false,
    "datastore_contains_all_records_of_source_file": false,
    "description": "Cantidad de cruceristas por temporada, puerto y nombre del crucero",
    "fileName": "cruceristas-por-puerto-crucero-temporada.csv",
    "format": "CSV",
    "hash": "",
    "id": "turismo_f7ce925a-0a48-4e38-8f48-360d6513fd83",
    "last_modified": "2023-09-29T14:43:19.057358",
    "metadata_modified": "2023-11-07T11:26:55.764273",
    "mimetype": null,
    "mimetype_inner": null,
    "name": "Cruceristas por temporada, puerto y crucero",
    "package_id": "turismo_fbc269ea-5f71-45b6-b70c-8eb38a03b8db",
    "position": 0,
    "resource_type": null,
    "size": null,
    "state": "active",
    "url": "http://datos.yvera.gob.ar/dataset/fbc269ea-5f71-45b6-b70c-8eb38a03b8db/resource/f7ce925a-0a48-4e38-8f48-360d6513fd83/download/cruceristas-por-puerto-crucero-temporada.csv",
    "url_type": null
  }
```

**Comment by amercader** [comment]
@pdelboca I don't recall the harvester extension changing the incoming resource ids in any way, but maybe I'm wrong. Or maybe you have other extensions/custom schemas at play?

In any case perhaps we can loosen up the requirement a bit by using an `uuid_if_not_a_sysadmin` validator but the consensus was that long term it made sense to enforce UUIDs for id fields for security reasons.

**Comment by ThrawnCA** [comment]
> * "Value is a valid UUID". In the past we've allowed different values here to support potential use cases but it's probably safer to enforce UUIDs for all ids and use custom fields if needed

How are sites with existing resources supposed to migrate them to UUIDs?

Due to the stricter validation, those resources cannot even be edited. How is legacy data supposed to be fixed?

**Comment by wardi** [comment]
@ThrawnCA validation on id fields can be overridden using custom schemas for sites that would like to continue using non-uuid values, at least until or unless we switch the models to use proper uuid columns.

**Comment by ThrawnCA** [comment]
> [@ThrawnCA](https://github.com/ThrawnCA) validation on id fields can be overridden using custom schemas for sites that would like to continue using non-uuid values, at least until or unless we switch the models to use proper uuid columns.

Yes, but if a site does want to switch to using UUIDs, how are they to convert existing data?

**Comment by wardi** [comment]
Reassigning resource ids is complicated by a few things:

- resource views
- any sort of cloud uploads
- datastore tables created by datapusher/xloader
- datastore-first tables created by API or table designer
- existing API integrations expecting stable resource ids

If I was running a site with custom resource ids I'd override the schema to allow people to keep using the ids they're familiar with, and to avoid dealing with these problems.

If forced to migrate I would do it with SQL directly in application and datastore databases, reload the search index and work with outside integrations to migrate their code to the new ids.

It's possible to migrate using the API e.g. with things like https://github.com/ckan/ckanapi/pull/229 but it would be much slower than a custom SQL migration.
